[ao] Added README for ModelReport functionality

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #81369
* #81382

Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Added README for ModelReport functionality

Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+## Implementing Your Own Detector
+
+The main way to add functionality to the ModelReport API is to add more Detectors.
+Detectors each have a specific focus in terms of the type of information they collect information on.
```
type of information they collect?

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+
+### Requirements to Implement Detector
+All Detectors inherit from the `DetectorBase` class, and all of them (including any custom detectors you create) will need to implement 3 methods:
+- `determine_observer_insert_points(self, model)` -> `Dict`: determines which observers you want into insert into a model to gather statistics and where in the model. All of them return a dictionary mapping unique observer fully qualified names (fqns), which is where we want to insert them, to a dictionary of location and argument information in the format:
```
```suggestion
- `determine_observer_insert_points(self, model)` -> `Dict`: determines which observers you want to insert into a model to gather statistics and where in the model. All of them return a dictionary mapping unique observer fully qualified names (fqns), which is where we want to insert them, to a dictionary of location and argument information in the format:
```

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+    "[unique_observer_fqn_of_insert_location]" :
+    {
+        "target_node" -> the node we are trying to observe with this observer (torch.fx.node.Node),
+        "insert_observer" -> the observer we wish to insert (ObserverBase),
```
I feel `insert_observer` is not very intuitive, maybe rename this to `observer_to_insert` or something similar?

vspenubarthi replied:
Addressed in new downstream commit here: https://github.com/pytorch/pytorch/pull/81382

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+    {
+        "target_node" -> the node we are trying to observe with this observer (torch.fx.node.Node),
+        "insert_observer" -> the observer we wish to insert (ObserverBase),
+        "insert_post" -> True if this is meant to be a post-observer for target_node, False if pre-observer,
```
typically bool will starts with `is_`, maybe `is_post_observer`?

vspenubarthi replied:
Addressed in new downstream commit here: https://github.com/pytorch/pytorch/pull/81382

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+  - A string-based report that is easily digestable and actionable explaining the data collected by relavent observers for that detector
+  - A dictionary containing statistics collected by the relavent observers and values calculated by the detector for futher analysis or plotting
+
+
```
should you talk about `ModelReporterObserver` and it's relationship with detectors as well

jerryzh168 commented on torch/ao/quantization/fx/_model_report/README.md:
```diff
@@ -0,0 +1,119 @@
+Meanwhile, the `InputWeightEqualizationDetector` determines whether Input-Weight Equalization should be applied for each layer.
+
+
+### Requirements to Implement Detector
```
I feel the design for detector should be close to model reporter design overview. and also we can talk about high level flow and usage before talking about design details

jerryzh168 reviewed (approved):
looks good overall, had a few comments inline

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Added README for ModelReport functionality"


Summary: This adds a README for the ModelReport functionality that
contains an overview of the class, what it does,
and how it works, an example of usage, information on how to implement a
new detector (since this is how core functionality is added), folder
structure information, and finally information on tests and where they
are located.

The ModelReport class is still in development and will, in the future,
get additional features such as visualizations, and the README will be
updated with this information as it is added.

Test Plan: Just a new README, no code is added, README will be reviewed
for accuracy and ease of use/ easiness to read.

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
